### ISSUE TITLE
Fails to copy non-resident files under 1 cluster size.

### ISSUE THREAD

**Issue by Aaron-Ritter-IDEXX** [dr-anoroc/rawccopy#2]
I'd try it with the original, but the security software on my dev PC won't let it run. Imagine that, haha.

I'm not very experienced with NTFS or with C, so please bear with me while I walk through what I've found.

I believe that non-resident files of less than 1 full cluster all fail to copy.

The difference seems to be at processor.c line 209:
`if (!start_at->non_resident)`
In the files that cannot be copied, non_resident is true.

Later on at line 220, TO_CLUST(context,bt_cnt) is called, which **appears** to calculate the expected number of clusters, and if it's 0, the extraction of this data stream is skipped. But the thing is, for a file with fewer bytes than one cluster, the result of this calculation will incorrectly be 0. At line 227, the total clusters are again compared to this value, which reinforces my impression that it's supposed to be the expected number of clusters.

My conclusion is that I believe the TO_CLUST macro has an error. Instead of
`#define TO_CLUST(context,bt_cnt) ((bt_cnt) / BytesPerCluster(context->boot))`
you should want
`#define TO_CLUST(context,bt_cnt) (bt_cnt > 0 ? 1 + (bt_cnt - 1) / BytesPerCluster(context->boot) : 0)`
This alteration seems to fix my use cases.

Now, a non-resident, 0-byte file will simply not have the file created at all. That's no good, so an edit at line 210:
[redacted for being incorrect; see next comment]

**Comment by Aaron-Ritter-IDEXX** [comment]
I seem to have introduced a new bug, so don't celebrate yet. I suspect the real_sz isn't what I thought it was.

edit: Indeed, I guess real_sz can be 0 for resident files of non-zero size. Then the correct code block at 210 looks like this:

```
	if (!start_at->non_resident)
	{
		bytes payload = GetAttributeData(context->cr, start_at);
		result = WriteDataToDestination(context, stream_name, payload);
		DeleteBytes(payload);
		return result;
	}
	else if (start_at->real_sz == 0)
	{
		bytes buf = ZeroBuffer(0);
		WriteDataToDestination(context, stream_name, buf);
		return result;
	}
```

**Comment by Aaron-Ritter-IDEXX** [comment]
I created this pull request after finishing all of my testing. https://github.com/dr-anoroc/rawccopy/pull/3

**Comment by dr-anoroc** [comment]
Hey thanks for picking this up. I was able to reproduce it indeed.

I have a couple of comments on your proposed fix however. 

First of all, there is a minor memory leak in the second half of your code:

``` 
else if (start_at->real_sz == 0)
{
	bytes buf = ZeroBuffer(0);
	WriteDataToDestination(context, stream_name, buf);
	return result;
}
```
The memory in `buf` is never freed. This is not a big deal really, as it happens right before the program exits, but still. In any case, this could be fixed as follows:
``` 
else if (start_at->real_sz == 0)
{
	bytes buf = ZeroBuffer(0);
	result = WriteDataToDestination(context, stream_name, buf);
	DeleteBytes(buf);
	return result;
}
```
Looking closer at the problem you were trying to fix by this second branch of the `if` however, it struck me that there was a more fundamental flaw in the logic: this whole first part of the code (up until line 234 in the original file) shouldn't be using a "cluster-logic" at all. In fact the `TO_CLUST` macro's behavior is not well-defined for any value of `bt_cnt` that is not a multiple of the cluster size.

So the real conclusion here is that the initial part of the function should be rewritten to use the byte count instead of the cluster count. That removes the need to adapt the `TO_CLUST` macro and also the additional `else if ... ` is no longer necessary.

While hunting for a file to reproduce the second case (non-resident data attribute, but zero size), I stumbled on another bug. 
In conclusion, what I'm going to do is reopen this issue and create a new one for the bug I discovered and then commit a fix for both of them in the coming days.

**Comment by Aaron-Ritter-IDEXX** [comment]
To the best of my understanding, `TO_CLUST` does correctly get the number of clusters an attribute occupies (rounding up to the nearest whole number), but the purpose of its use appears to be checking for the "incomplete attribute" situation, and I don't know what that could mean, due to my inexperience. In any case, you seem to have a handle on it now. I have no complaints about your plan.

I got a non-resident, 0-byte stream by copying a large (multi-cluster) file and then truncating it with trunc.exe. That can also easily get you non-resident files with byte count of less than 1 cluster or exactly a multiple. ( https://code.google.com/archive/p/trunc/ or https://www.softpedia.com/get/System/System-Miscellaneous/Trunc.shtml )

**Comment by dr-anoroc** [comment]
Fixed in 0.1.7
